src/quartz/simulator/schedule.cpp

```diff
@@ -1006,11 +1004,52 @@ get_schedules(const CircuitSeq &sequence,
   std::unordered_map<std::string, std::queue<int>> gate_indices;
 
   for (auto &local_qubit : local_qubits) {
-    // convert vector<int> to vector<bool>
-    std::vector<bool> local_qubit_mask(num_qubits, false);
+    auto previous_local_qubit_mask = local_qubit_mask;
+    // Convert vector<int> to vector<bool>.
+    local_qubit_mask.assign(num_qubits, false);
     for (auto &i : local_qubit) {
       local_qubit_mask[i] = true;
     }
+    std::cout << std::endl;
+    if (current_local_qubit_layout.empty()) {
+      // First iteration. Take the initial layout from |local_qubits[0]|.
+      current_local_qubit_layout = local_qubit;
+    } else {
+      // Update the layout.
+      // We should have the same number of local qubits.
+      assert(local_qubit.size() == current_local_qubit_layout.size());
+      int num_global_swaps = 0;
+      for (auto &i : local_qubit) {
+        if (!previous_local_qubit_mask[i]) {
+          num_global_swaps++;
+        }
+      }
+      int should_swap_ptr = 0;
+      for (int i = (int)current_local_qubit_layout.size() - num_global_swaps;
+           i < (int)current_local_qubit_layout.size(); i++) {
+        if (!local_qubit_mask[current_local_qubit_layout[i]]) {
+          // This qubit should be swapped with a global qubit,
+          // and it is already at the correct position.
+          continue;
+        }
+        while (local_qubit_mask[current_local_qubit_layout[should_swap_ptr]]) {
+          should_swap_ptr++;
+        }
+        // Find the first local qubit that should be swapped with a global
+        // qubit, and perform a local swap.
+        std::swap(current_local_qubit_layout[i],
+                  current_local_qubit_layout[should_swap_ptr]);
+        should_swap_ptr++;
+      }
+      for (auto &i : local_qubit) {
```

Comment:
I forgot to describe the global swap part, it does something like: for pos in [n_local, ... n_global-1]: if !local_qubit_mask[layout[pos]]: swap(layout[pos], layout[n_local-num_global_swaps]), num_global_swaps--.